### PR TITLE
kubeadm: Reduce the usage of InitConfiguration

### PR DESCRIPTION
**Pull request by rosti**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

Back in the `v1alpha1` we had a single flat `MasterConfiguration` and whenever some function needed something from it, it took a pointer to `MasterConfiguration` argument. In `v1beta1` this configuration is no longer flat, but composite with `InitConfiguration` bearing the closest resemblance to the old `MasterConfiguration`. Most of kubeadm still works with `InitConfiguration` even though most functions need only a portion or even a single field from the config.
This change strives to reduce the unnecessary usage of InitConfiguration. The approach taken here is to switch APIs, that take InitConfiguration to some other type of argument (usually ClusterConfiguration).

**Which issue(s) this PR fixes**:

Refs kubernetes/kubeadm#1084

**Special notes for your reviewer**:

This PR is marked as a WIP as I want some feedback on the scope of changes and weather it is too broad.
The commits aren't squashed yet, as it would be easier to drop something upon feedback if needed. Also, it might be a lot easier to review this on a per commit basis (rather than as a whole).

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/assign @fabriziopandini
/assign @timothysc

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by neolit123 — APPROVED** [review]
thanks @rosti
did a pass over the diff and the change LGTM.

/lgtm
/assign @fabriziopandini 
for approve.

**Review by yagonobre — APPROVED** [review]
LGTM
thanks @rosti

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Is there a reason for not switching  `InitConfiguration` into `ClusterConfiguration`\\`APIEndpoint` pair also here? (e.g. similar to what proposed for `getAPIServerCommand`)

File: cmd/kubeadm/app/util/endpoint.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/util/endpoint.go]: s/`cfg.API`/`localEndpoint `

File: cmd/kubeadm/app/util/endpoint.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/util/endpoint.go]: s/`api`/`localEndpoint`

File: cmd/kubeadm/app/util/endpoint.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/util/endpoint.go]: What do you think about replacing `cfg ClusterConfiguration` with `ControlPlaneEndpoint  string`?`

File: cmd/kubeadm/app/phases/etcd/local.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/phases/etcd/local.go]: what about changing the `getEtcdCommand` signature similarly to other `get*commands`?

Comment (fabriziopandini) [reply cmd/kubeadm/app/phases/etcd/local.go]: NB. if I'm not wrong, this unblock changing also the signature for `GetEtcdPodSpec` and `CreateStackedEtcdStaticPodManifestFile`

File: cmd/kubeadm/app/preflight/checks.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/preflight/checks.go]: what about changing `ClusterConfiguration` to `KubeProxy` configuration only?

**Review by fabriziopandini — COMMENTED** [review]
@rosti 💯for this cleanup!
I have left some minor comments but this IMO can merge as soon as you remove WIP.

File: cmd/kubeadm/app/phases/kubelet/config.go
Comment (rosti) [reply cmd/kubeadm/app/phases/kubelet/config.go]: Yep, I like that idea. But might, as well, change the name of the func to the more descriptive `CreateKubeletConfigMap`.

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (rosti) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Must have missed that one. Thanks for pointing it out!

File: cmd/kubeadm/app/preflight/checks.go
Comment (rosti) [reply cmd/kubeadm/app/preflight/checks.go]: We have a couple of options for this one:
1. Keep the func named `RunOptionalJoinNodeChecks` and use `ClusterConfiguration`.
2. Rename the func to `RunOptionalKubeProxyChecks` and pass `KubeProxy` configuration.

I don't think it's expected and acceptable for the func to be named `RunOptionalJoinNodeChecks` and to accept `KubeProxy` configuration directly. This is unexpected for the reader of the code and is bound to raise questions.

**Comment by rosti** [comment]
One of the commits (a9fd78dc49a) can probably be moved into a separate PR. Though I am not sure about this and it probably belongs here.

File: cmd/kubeadm/app/preflight/checks.go
Comment (neolit123) [reply cmd/kubeadm/app/preflight/checks.go]: we can keep this as is, as we might decide to add more checks in the function.
hopefully not, though.

**Comment by timothysc** [comment]
@rosti - please squash your commits.

**Comment by rosti** [comment]
@timothysc will do it, but just before I pull together the final version of things. Waiting for @fabriziopandini to review it.

**Comment by fabriziopandini** [comment]
@rosti green light for me! let's get this in as soon as possible
/approve

**Comment by rosti** [comment]
Commits squashed, WIP tag removed, this is finally ready.

**Pull request merged**
